modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
 
-  pragma "default string value" extern var defaultStringValue: string = noinit;
+  pragma "insert line file info"
+  extern proc chpl_mem_calloc(number: size_t, size: size_t,
+                              description: chpl_mem_descInt_t) : c_void_ptr;
 
-  // The following method is called by the compiler to determine the default
-  // value of a given type.  For strings, this should return the value
-  // defined above.
-  // strings will be records, so param is not possible for them
-  inline proc _defaultOf(type t) where t: string return defaultStringValue;
+  pragma "insert line file info"
+  extern proc chpl_mem_realloc(ptr: c_ptr, size: size_t,
+                               description: chpl_mem_descInt_t) : c_void_ptr;
 
-  // String concatenation
-  inline proc +(s: string, x: string) {
-    var cs = __primitive("string_concat", s.c_str(), x.c_str());
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  pragma "insert line file info"
+  extern proc chpl_mem_free(ptr: c_ptr): void;
+
+  // TODO: define my own mem descriptors?
+  extern const CHPL_RT_MD_STR_COPY_REMOTE: chpl_mem_descInt_t;
+  extern const CHPL_RT_MD_STR_COPY_DATA: chpl_mem_descInt_t;
+
+  inline proc chpl_string_comm_get(dest: bufferType, src_loc_id: int(64),
+                                   src_addr: bufferType, len: size_t) {
+    __primitive("chpl_comm_get", dest, src_loc_id, src_addr, len);
   }
 
-  inline proc +(s: c_string, x: string) {
-    var cs = __primitive("string_concat", s, x.c_str());
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  proc copyRemoteBuffer(src_loc_id: int(64), src_addr: bufferType, len: int): bufferType {
+      const dest = chpl_mem_alloc((len+1).safeCast(size_t), CHPL_RT_MD_STR_COPY_REMOTE): bufferType;
+      chpl_string_comm_get(dest, src_loc_id, src_addr, len.safeCast(size_t));
+      dest[len] = 0;
+      return dest;
   }
 
-  inline proc +(s: string, x: c_string) {
-    var cs = __primitive("string_concat", s.c_str(), x);
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  extern proc memmove(destination: c_ptr, const source: c_ptr, num: size_t);
+
+  // pointer arithmetic used for strings
+  inline proc +(a: c_ptr, b: integral) return __primitive("+", a, b);
+
+  config param debugStrings = false;
+}
+
+// Chapel Strings
+module String {
+  use BaseStringType;
+  use StringCasts;
+
+  inline proc chpl_debug_string_print(s: c_string) {
+    if debugStrings {
+      extern proc printf(format: c_string, x...);
+      printf("%s\n", s);
+    }
   }
 
-  inline proc +(s: string, x: numeric)
-    return s + x:string;
+  // TODO: We should be able to remove "ignore noinit", but doing so causes
+  // memory leaks in various places. Investigate why later and add noinit back
+  // in when possible.
```

Comment:
Without ignore noinit, some code in normalize will prevent a string record from ever be correctly returned, because it no-inits the LHS of an assignment call. I personally think we should do two things:
 1) make the compiler not add calls to assignment in these situations (just init_copy or equivalent)
 2) automatically add "ignore noinit" to records that are not POD (e.g. have overloaded assignment).
So, I think record string will always need "ignore noinit" within the compiler. I just think the compiler
should be adding that flag automatically.

It is less obvious to me why you have specified "no default functions".